bin/reflective_loader/src/loader.rs

```diff
@@ -0,0 +1,844 @@
+            loop {
+                // Simply dereferencing a pointer may result in the struct being copied instead of referenced.
+                // let mut library_thunk: IMAGE_THUNK_DATA64 = unsafe { *library_thunk_ref };
+                // Instead we need to dereference to a mutable reference.
+                // We can't just set it equal since that will be a pointer to the object.
+                // To use it each line would need to dereference the pointer then access the field.
+                // let mut library_thunk: *mut IMAGE_THUNK_DATA64 = library_thunk_ref;
+                let mut library_thunk = unsafe { &mut *library_thunk_ptr };
+
+                // Access of a union field is unsafe
+                if unsafe{library_thunk.u1.AddressOfData} == 0 {
+                    break;
+                }
+                if image_snap_by_ordinal(unsafe{library_thunk.u1.Ordinal as usize}) {
+                    // Calculate the ordinal reference to the function from the library_thunk entry.
+                    let function_ordinal_ptr = image_ordinal(unsafe{library_thunk.u1.Ordinal as usize}) as *const u8;
+                    // Get the address of the function using `GetProcAddress` and update the thunks reference.
+                    library_thunk.u1.Function = get_proc_address(get_proc_address_fn, get_last_error_fn ,library_handle, function_ordinal_ptr) as _;
+                } else {
+                    // Calculate a refernce to the function name by adding the dll_base and name's RVA.
+                    let image_import_ptr: *mut IMAGE_IMPORT_BY_NAME = (new_dll_base as usize + unsafe{library_thunk.u1.AddressOfData} as usize) as *mut IMAGE_IMPORT_BY_NAME;
+                    // Get the address of the function using `GetProcAddress` and update the thunks reference.
+                    let function_name_ptr = unsafe {(*image_import_ptr ).Name.as_ptr()};
+                    // let tmp_new_func_addr = unsafe{ get_proc_address_fn(library_handle, function_name).unwrap() as _};
+                    let tmp_new_func_addr = get_proc_address(get_proc_address_fn, get_last_error_fn, library_handle, function_name_ptr) as _;
+                    library_thunk.u1.Function = tmp_new_func_addr;
+                }
+                library_thunk_ptr = (library_thunk_ptr as usize + core::mem::size_of::<usize>()) as *mut IMAGE_THUNK_DATA64;
+            }
+        }
+        base_image_import_table = (base_image_import_table as usize + core::mem::size_of::<IMAGE_IMPORT_DESCRIPTOR>() as usize) as *mut IMAGE_IMPORT_DESCRIPTOR;
+    }
+
+}
+
+/// Get a pointer to the Thread Environment Block (TEB)
+pub unsafe fn get_teb() -> *mut ntapi::ntpebteb::TEB 
+{
+    let teb: *mut ntapi::ntpebteb::TEB;
+    asm!("mov {teb}, gs:[0x30]", teb = out(reg) teb);
+    teb
+}
+
+/// Get a pointer to the Process Environment Block (PEB)
+pub unsafe fn get_peb() -> *mut PEB 
```

Comment:
`gs:[0x60]` is the way you would usually get the peb

Reply:
Ooooh very dope tyvm.